toolchain/diagnostics/diagnostic_emitter.h

```diff
@@ -101,10 +89,20 @@ struct Diagnostic {
   llvm::SmallVector<llvm::Any, 0> format_args;
 
   // Returns the formatted string. By default, this uses llvm::formatv.
-  std::function<std::string(const Diagnostic&)> format_fn;
+  std::function<std::string(const DiagnosticMessage&)> format_fn;
+};
+
+// An instance of a single error or warning.  Information about the diagnostic
+// can be recorded into it for more complex consumers.
+struct Diagnostic {
+  // The diagnostic's level.
+  DiagnosticLevel level;
+
+  // The main error or warning.
+  DiagnosticMessage main;
```

Comment:
What do you think about naming this `message` like so:

```suggestion
  // The main error or warning message to use.
  DiagnosticMessage message;
```

Seems a bit better to me than `main`.

Reply:
Switching, but for me `message` is synonymous with `notes` so it makes this a pretty confusing struct.